src/lib/runtime/files/11_udsep/11_udsep.c

```diff
@@ -0,0 +1,97 @@
+/* 
+ * Copyright (c) 2017-2018, SyLabs, Inc. All rights reserved.
+ * Copyright (c) 2017, SingularityWare, LLC. All rights reserved.
+ *
+ * Copyright (c) 2015-2017, Gregory M. Kurtzer. All rights reserved.
+ * 
+ * Copyright (c) 2016-2017, The Regents of the University of California,
+ * through Lawrence Berkeley National Laboratory (subject to receipt of any
+ * required approvals from the U.S. Dept. of Energy).  All rights reserved.
+ * 
+ * This software is licensed under a customized 3-clause BSD license.  Please
+ * consult LICENSE file distributed with the sources of this project regarding
+ * your rights to use or distribute this software.
+ * 
+ * NOTICE.  This Software was developed under funding from the U.S. Department of
+ * Energy and the U.S. Government consequently retains certain rights. As such,
+ * the U.S. Government has been granted for itself and others acting on its
+ * behalf a paid-up, nonexclusive, irrevocable, worldwide license in the Software
+ * to reproduce, distribute copies to the public, prepare derivative works, and
+ * perform publicly and display publicly, and to permit other to do so. 
+ * 
+*/
+
+#include <errno.h>
+#include <fcntl.h>
+#include <stdio.h>
+#include <string.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <limits.h>
+#include <unistd.h>
+#include <stdlib.h>
+#include <grp.h>
+#include <pwd.h>
+
+#include "config.h"
+#include "util/file.h"
+#include "util/util.h"
+#include "util/config_parser.h"
+#include "util/message.h"
+#include "util/privilege.h"
+#include "util/registry.h"
+
+#include "../file-bind.h"
+#include "../../runtime.h"
+
+
+int _singularity_runtime_files_11_user_defined_SINGULARITYENV_PATH(void) {
+    char *tmp_file;
+    char *containerdir = CONTAINER_FINALDIR;
+    char *tmpdir = singularity_registry_get("SESSIONDIR");
+    char *user_add_path = singularity_registry_get("USER_DEFINED_PREPEND");
+
+    singularity_message(DEBUG, "Called _singularity_runtime_files_11-user_defined_SINGULARITYENV_PATH()\n");
+
+    if ( user_add_path == NULL ) {
+        singularity_message(VERBOSE2, "No user defined SINGULARITYENV_PATH found.\n");
+        return 0; 
+    }
+
+    if ( tmpdir == NULL ) {
+        singularity_message(ERROR, "Failed to obtain session directory\n");
+        ABORT(255);
+    }
+
+    tmp_file = joinpath(tmpdir, "/11-user_defined_SINGULARITYENV_PATH.sh");
+
+    singularity_message(VERBOSE2, "Creating empty /.singularity.d/env/11-user_defined_SINGULARITYENV_PATH.sh in %s\n", containerdir);
+    singularity_priv_escalate();
+    if ( ( fileput(joinpath(containerdir, "/.singularity.d/env/11-user_defined_SINGULARITYENV_PATH.sh"), "") ) !=0 ) {
```

Comment:
If overlayfs is not supported or not enabled in singularity.conf and if image is not writable, test will always fail